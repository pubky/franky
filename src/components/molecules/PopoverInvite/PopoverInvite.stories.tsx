import type { Meta, StoryObj } from '@storybook/nextjs-vite';

import { PopoverInvite } from './PopoverInvite';

const meta: Meta<typeof PopoverInvite> = {
  title: 'Molecules/Popover/PopoverInvite',
  component: PopoverInvite,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: 'Invite popover providing quick links to Pubky contact channels.',
      },
    },
  },
  args: {
    className: 'hover:bg-brand/10',
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Playground: Story = {
  render: (args) => <PopoverInvite {...args} />,
};

export const CustomTriggerStyle: Story = {
  args: {
    className: 'hover:bg-secondary/20 text-secondary-foreground',
  },
  parameters: {
    docs: {
      description: {
        story: 'Example showing how to adapt the trigger button styling using className.',
      },
    },
  },
};
