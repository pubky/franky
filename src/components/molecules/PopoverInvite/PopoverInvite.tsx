import { Gift, Mail, Twitter, Send } from 'lucide-react';

import * as Atoms from '@/atoms';
import * as Molecules from '@/molecules';

interface PopoverInviteProps {
  className?: string;
}

// TODO: extract to env
const EMAIL_URL = 'mailto:hello@pubky.com';
const TWITTER_URL = 'https://x.com/pubky';
const TELEGRAM_URL = 'https://t.me/pubky';

export function PopoverInvite({ className = 'hover:bg-brand/10' }: PopoverInviteProps) {
  return (
    <Molecules.Popover>
      <Molecules.PopoverTrigger asChild>
        <Atoms.Button variant="ghost" size="icon" className={className}>
          <Gift className="h-4 w-4 text-brand" />
        </Atoms.Button>
      </Molecules.PopoverTrigger>
      <Molecules.PopoverContent className="w-[327px]">
        <Atoms.Container className="flex-col gap-6 px-3 py-2">
          <Atoms.Container className="flex-col gap-2">
            <Atoms.Heading level={4} size="sm" className="text-popover-foreground">
              Don&apos;t have an invite yet?
            </Atoms.Heading>
            <Atoms.Typography size="sm" className="text-muted-foreground">
              Ask the Pubky team!
            </Atoms.Typography>
          </Atoms.Container>
          <Atoms.Container className="flex-row gap-4">
            <Atoms.Link href={EMAIL_URL}>
              <Mail className="h-6 w-6" />
            </Atoms.Link>
            <Atoms.Link href={TWITTER_URL}>
              <Twitter className="h-6 w-6" />
            </Atoms.Link>
            <Atoms.Link href={TELEGRAM_URL}>
              <Send className="h-6 w-6" />
            </Atoms.Link>
          </Atoms.Container>
        </Atoms.Container>
      </Molecules.PopoverContent>
    </Molecules.Popover>
  );
}
