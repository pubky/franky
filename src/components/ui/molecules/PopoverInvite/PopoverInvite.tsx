import { Gift, Mail, Twitter, Send } from 'lucide-react';
import { Button, Link, Popover, PopoverContent, PopoverTrigger, Heading, Container, Typography } from '@/components/ui';

interface PopoverInviteProps {
  className?: string;
}

// TODO: extract to env
const EMAIL_URL = 'mailto:hello@pubky.com';
const TWITTER_URL = 'https://x.com/pubky';
const TELEGRAM_URL = 'https://t.me/pubky';

export function PopoverInvite({ className = 'hover:bg-brand/10' }: PopoverInviteProps) {
  return (
    <Popover>
      <PopoverTrigger asChild>
        <Button variant="ghost" size="icon" className={className}>
          <Gift className="h-4 w-4 text-brand" />
        </Button>
      </PopoverTrigger>
      <PopoverContent className="w-[327px]">
        <Container className="flex-col gap-6 px-3 py-2">
          <Container className="flex-col gap-2">
            <Heading level={4} size="sm" className="text-popover-foreground">
              Don&apos;t have an invite yet?
            </Heading>
            <Typography size="sm" className="text-muted-foreground">
              Ask the Pubky team!
            </Typography>
          </Container>
          <Container className="flex-row gap-4">
            <Link href={EMAIL_URL}>
              <Mail className="h-6 w-6" />
            </Link>
            <Link href={TWITTER_URL}>
              <Twitter className="h-6 w-6" />
            </Link>
            <Link href={TELEGRAM_URL}>
              <Send className="h-6 w-6" />
            </Link>
          </Container>
        </Container>
      </PopoverContent>
    </Popover>
  );
}
