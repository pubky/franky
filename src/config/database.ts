import { Env } from '@/libs';

export const DB_NAME = Env.NEXT_PUBLIC_DB_NAME;
export const DB_VERSION = Env.NEXT_PUBLIC_DB_VERSION;
