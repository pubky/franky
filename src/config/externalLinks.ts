import { Env } from '@/libs';

/**
 * External Links Configuration
 *
 * This file centralizes all external URLs used throughout the application.
 * URLs can be overridden using environment variables for different environments.
 */

// Pubky ecosystem links
export const PUBKY_RING_URL = Env.NEXT_PUBLIC_PUBKY_RING_URL;
export const PUBKY_CORE_URL = Env.NEXT_PUBLIC_PUBKY_CORE_URL;

// Social media links
export const TWITTER_URL = Env.NEXT_PUBLIC_TWITTER_URL;
export const TWITTER_GETPUBKY_URL = Env.NEXT_PUBLIC_TWITTER_GETPUBKY_URL;
export const TELEGRAM_URL = Env.NEXT_PUBLIC_TELEGRAM_URL;
export const GITHUB_URL = Env.NEXT_PUBLIC_GITHUB_URL;

// Contact links
export const EMAIL_URL = `mailto:${Env.NEXT_PUBLIC_EMAIL}`;

// App store links
export const APP_STORE_URL = Env.NEXT_PUBLIC_APP_STORE_URL;
export const PLAY_STORE_URL = Env.NEXT_PUBLIC_PLAY_STORE_URL;
