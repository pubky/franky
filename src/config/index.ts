export * from './database';
export * from './network';
export * from './nexus';
export * from './sync';
export * from './logs';
export * from './metadata';
