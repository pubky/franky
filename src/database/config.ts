import { logger } from '@/lib/logger';

// Get database version from environment variable
export const DB_VERSION = Number(process.env.NEXT_PUBLIC_VERSION_DB) || 1;

if (!process.env.NEXT_PUBLIC_VERSION_DB) {
  logger.warn('NEXT_PUBLIC_VERSION_DB is not defined in .env file. Using default version 1');
}

export const DB_NAME = 'FrankyDB'; 