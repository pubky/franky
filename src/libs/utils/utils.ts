import { clsx, type ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

type ExtractInitialsProps = { name: string; maxLength: number };
type CopyToClipboardProps = { text: string };
type FormatPublicKeyProps = { key: string; length: number };

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function formatPublicKey({ key, length = 12 }: FormatPublicKeyProps) {
  if (!key) return '';
  if (key.length <= length) return key;
  const prefix = key.slice(0, length / 2);
  const suffix = key.slice(-length / 2);
  return `${prefix}...${suffix}`;
}

export async function copyToClipboard({ text }: CopyToClipboardProps) {
  if (!navigator.clipboard) {
    throw new Error('Clipboard API not supported');
  }

  await navigator.clipboard.writeText(text);
}

// Helper function to normalize Radix UI IDs in container HTML for snapshot tests
export const normaliseRadixIds = (container: HTMLElement) => {
  const clonedContainer = container.cloneNode(true) as HTMLElement;
  const elementsWithAriaControls = clonedContainer.querySelectorAll('[aria-controls]');
  elementsWithAriaControls.forEach((el) => {
    const ariaControls = el.getAttribute('aria-controls');
    if (ariaControls && ariaControls.includes('radix-«r')) {
      el.setAttribute('aria-controls', 'radix-«r0»');
    }
  });
  return clonedContainer;
};

const customCases = [
  { name: 'bitcoin', color: '#FF9900' },
  { name: 'synonym', color: '#FF6600' },
  { name: 'bitkit', color: '#FF4400' },
  { name: 'pubky', color: '#C8FF00' },
  { name: 'blocktank', color: '#FFAE00' },
  { name: 'tether', color: '#26A17B' },
];

/**
 * Generates a consistent color for a given string
 * @param str - The string to generate a color for
 * @returns Hex color string
 */
export function generateRandomColor(str: string): string {
  const lowerStr = str.toLowerCase();

  // Check for cases
  for (const special of customCases) {
    if (lowerStr === special.name) {
      return special.color;
    }
  }

  // Generate a hash value from the input string
  const hash = Array.from(str).reduce((hash, char) => {
    return char.charCodeAt(0) + ((hash << 5) - hash);
  }, 0);

  // Ensure the hash is non-negative
  const positiveHash = Math.abs(hash);

  // Convert hash to a 2-character hex value
  const randomByte = positiveHash & 0xff; // extract the lowest 8 bits
  const randomHex = randomByte.toString(16).padStart(2, '0');

  // Pick a random pattern
  const patterns = [
    `FF00${randomHex}`,
    `FF${randomHex}00`,
    `${randomHex}FF00`,
    `${randomHex}00FF`,
    `00${randomHex}FF`,
    `00FF${randomHex}`,
  ];

  // Select pattern based on the hash
  const pattern = patterns[positiveHash % patterns.length];

  return `#${pattern}`;
}

/**
 * Converts a hex color to rgba format
 * @param hex - Hex color string (e.g., '#FF9900')
 * @param alpha - Alpha value (0-1)
 * @returns RGBA color string
 */
export function hexToRgba(hex: string, alpha: number) {
  const [r, g, b] = hex.match(/\w\w/g)!.map((x) => parseInt(x, 16));
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
}

export function extractInitials({ name, maxLength = 2 }: ExtractInitialsProps) {
  if (!name || typeof name !== 'string') return '';

  return name
    .trim()
    .split(/\s+/)
    .filter((word) => word.length > 0)
    .map((word) => word.charAt(0))
    .join('')
    .toUpperCase()
    .slice(0, maxLength);
}

export function formatInviteCode(code: string) {
  if (!code) return '';

  // Remove all non-alphanumeric characters
  const cleaned = code.replace(/[^a-zA-Z0-9]/g, '');

  // Convert to uppercase
  const uppercased = cleaned.toUpperCase();

  // If no valid characters, return empty string
  if (uppercased.length === 0) return '';

  // Handle different lengths
  if (uppercased.length <= 4) {
    return uppercased;
  } else if (uppercased.length <= 8) {
    return `${uppercased.slice(0, 4)}-${uppercased.slice(4)}`;
  } else {
    // Take only first 12 characters for longer codes
    const truncated = uppercased.slice(0, 12);
    return `${truncated.slice(0, 4)}-${truncated.slice(4, 8)}-${truncated.slice(8, 12)}`;
  }
}

export function clearCookies() {
  if (typeof document !== 'undefined') {
    document.cookie.split(';').forEach((cookie) => {
      const eqPos = cookie.indexOf('=');
      const name = eqPos > -1 ? cookie.substr(0, eqPos) : cookie;
      document.cookie = `${name.trim()}=;expires=Thu, 01 Jan 1970 00:00:00 GMT;path=/`;
    });
  }
}
