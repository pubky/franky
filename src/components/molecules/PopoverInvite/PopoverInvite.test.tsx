import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { PopoverInvite } from './PopoverInvite';
import { EMAIL_URL, TWITTER_URL, TELEGRAM_URL } from '@/config';
import { normaliseRadixIds } from '@/libs/utils/utils';

// Mock @/libs to intercept all icons and utilities
vi.mock('@/libs', () => ({
  cn: (...args: unknown[]) => args.filter(Boolean).join(' '),
  Gift: ({ className }: { className?: string }) => (
    <div data-testid="gift-icon" className={className}>
      Gift
    </div>
  ),
  Mail: ({ className }: { className?: string }) => (
    <div data-testid="mail-icon" className={className}>
      Mail
    </div>
  ),
  XTwitter: ({ className }: { className?: string }) => (
    <div data-testid="xtwitter-icon" className={className}>
      XTwitter
    </div>
  ),
  Telegram: ({ className }: { className?: string }) => (
    <div data-testid="telegram-icon" className={className}>
      Telegram
    </div>
  ),
}));

describe('InvitePopover', () => {
  it('renders trigger button with gift icon', () => {
    render(<PopoverInvite />);

    const button = screen.getByRole('button');
    const giftIcon = screen.getByTestId('gift-icon');

    expect(button).toBeInTheDocument();
    expect(giftIcon).toBeInTheDocument();
    expect(giftIcon).toHaveClass('h-4', 'w-4', 'text-brand');
  });

  it('shows popover content when clicked', () => {
    render(<PopoverInvite />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(screen.getByText("Don't have an invite yet?")).toBeInTheDocument();
    expect(screen.getByText('Ask the Pubky team!')).toBeInTheDocument();
  });

  it('renders social contact links in popover', () => {
    render(<PopoverInvite />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    const mailIcon = screen.getByTestId('mail-icon');
    const twitterIcon = screen.getByTestId('xtwitter-icon');
    const telegramIcon = screen.getByTestId('telegram-icon');

    expect(mailIcon).toBeInTheDocument();
    expect(twitterIcon).toBeInTheDocument();
    expect(telegramIcon).toBeInTheDocument();

    // Check icons have proper classes
    expect(mailIcon).toHaveClass('h-6', 'w-6');
    expect(twitterIcon).toHaveClass('h-6', 'w-6');
    expect(telegramIcon).toHaveClass('h-6', 'w-6');
  });

  it('uses default URLs when no custom URLs provided', () => {
    render(<PopoverInvite />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    const mailLink = screen.getByTestId('mail-icon').parentElement;
    const twitterLink = screen.getByTestId('xtwitter-icon').parentElement;
    const telegramLink = screen.getByTestId('telegram-icon').parentElement;

    expect(mailLink).toHaveAttribute('href', EMAIL_URL);
    expect(twitterLink).toHaveAttribute('href', TWITTER_URL);
    expect(telegramLink).toHaveAttribute('href', TELEGRAM_URL);
  });

  it('uses default URLs', () => {
    render(<PopoverInvite />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    const mailLink = screen.getByTestId('mail-icon').parentElement;
    const twitterLink = screen.getByTestId('xtwitter-icon').parentElement;
    const telegramLink = screen.getByTestId('telegram-icon').parentElement;

    expect(mailLink).toHaveAttribute('href', EMAIL_URL);
    expect(twitterLink).toHaveAttribute('href', TWITTER_URL);
    expect(telegramLink).toHaveAttribute('href', TELEGRAM_URL);
  });

  it('applies custom className to trigger button', () => {
    render(<PopoverInvite className="custom-invite-class" />);

    const button = screen.getByRole('button');
    expect(button).toHaveClass('custom-invite-class');
  });

  it('has proper popover content structure', () => {
    render(<PopoverInvite />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    // Check the popover content structure
    const heading = screen.getByText("Don't have an invite yet?");
    const description = screen.getByText('Ask the Pubky team!');

    // Check that the components are rendered
    expect(heading).toBeInTheDocument();
    expect(heading).toHaveClass('text-popover-foreground');
    expect(description).toBeInTheDocument();
    expect(description).toHaveClass('font-normal', 'text-muted-foreground');
  });
});

describe('PopoverInvite - Snapshots', () => {
  it('matches snapshot for default PopoverInvite', () => {
    const { container } = render(<PopoverInvite />);
    const normalizedContainer = normaliseRadixIds(container);
    expect(normalizedContainer.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for PopoverInvite with custom className', () => {
    const { container } = render(<PopoverInvite className="custom-invite-style" />);
    const normalizedContainer = normaliseRadixIds(container);
    expect(normalizedContainer.firstChild).toMatchSnapshot();
  });
});
