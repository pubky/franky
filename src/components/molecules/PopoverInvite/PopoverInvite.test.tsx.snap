// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopoverInvite - Snapshots > matches snapshot for PopoverInvite with custom className 1`] = `
<button
  aria-controls="radix-«r0»"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 border-none size-9 custom-invite-style"
  data-as-child="true"
  data-size="icon"
  data-slot="popover-trigger"
  data-state="closed"
  data-testid="popover-trigger"
  data-variant="ghost"
  type="button"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-gift h-4 w-4 text-brand"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      height="4"
      rx="1"
      width="18"
      x="3"
      y="8"
    />
    <path
      d="M12 8v13"
    />
    <path
      d="M19 12v7a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2v-7"
    />
    <path
      d="M7.5 8a2.5 2.5 0 0 1 0-5A4.8 8 0 0 1 12 8a4.8 8 0 0 1 4.5-5 2.5 2.5 0 0 1 0 5"
    />
  </svg>
</button>
`;

exports[`PopoverInvite - Snapshots > matches snapshot for default PopoverInvite 1`] = `
<button
  aria-controls="radix-«r0»"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs hover:text-accent-foreground dark:hover:bg-accent/50 border-none size-9 hover:bg-brand/10"
  data-as-child="true"
  data-size="icon"
  data-slot="popover-trigger"
  data-state="closed"
  data-testid="popover-trigger"
  data-variant="ghost"
  type="button"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-gift h-4 w-4 text-brand"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      height="4"
      rx="1"
      width="18"
      x="3"
      y="8"
    />
    <path
      d="M12 8v13"
    />
    <path
      d="M19 12v7a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2v-7"
    />
    <path
      d="M7.5 8a2.5 2.5 0 0 1 0-5A4.8 8 0 0 1 12 8a4.8 8 0 0 1 4.5-5 2.5 2.5 0 0 1 0 5"
    />
  </svg>
</button>
`;
